Compile NestedTensor with AOTAutograd

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110529

This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

soulitzer committed:
AOTAutograd NestedTensor testing

[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"




[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"




[ghstack-poisoned]

soulitzer commented:
cc @bdhirsh

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"


Repro:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
nt, offsets = jagged_from_list([a, b, c], None)
nt2, _ = jagged_from_list([a, b, c], offsets)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```

Error:
```
(pytorch3) jw3468-mbp:pytorch3 jw3468$ t3
Traceback (most recent call last):
  File "/Users/jw3468/local/.tests/tst3.py", line 932, in <module>
    out = compiled_f(nt, nt2)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 451, in transform
    tracer.run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2103, in run
    super().run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 743, in run
    and self.step()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 706, in step
    getattr(self, inst.opname)(inst)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2213, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 855, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/jw3468/opt/miniconda3/envs/pytorch3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 983, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1050, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1035, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/__init__.py", line 1642, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4541, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"


Repro:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
nt, offsets = jagged_from_list([a, b, c], None)
nt2, _ = jagged_from_list([a, b, c], offsets)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```

Error:
```
(pytorch3) jw3468-mbp:pytorch3 jw3468$ t3
Traceback (most recent call last):
  File "/Users/jw3468/local/.tests/tst3.py", line 932, in <module>
    out = compiled_f(nt, nt2)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 451, in transform
    tracer.run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2103, in run
    super().run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 743, in run
    and self.step()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 706, in step
    getattr(self, inst.opname)(inst)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2213, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 855, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/jw3468/opt/miniconda3/envs/pytorch3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 983, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1050, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1035, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/__init__.py", line 1642, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4541, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"


Repro:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
nt, offsets = jagged_from_list([a, b, c], None)
nt2, _ = jagged_from_list([a, b, c], offsets)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```

Error:
```
(pytorch3) jw3468-mbp:pytorch3 jw3468$ t3
Traceback (most recent call last):
  File "/Users/jw3468/local/.tests/tst3.py", line 932, in <module>
    out = compiled_f(nt, nt2)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 451, in transform
    tracer.run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2103, in run
    super().run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 743, in run
    and self.step()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 706, in step
    getattr(self, inst.opname)(inst)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2213, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 855, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/jw3468/opt/miniconda3/envs/pytorch3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 983, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1050, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1035, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/__init__.py", line 1642, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4541, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"


Repro:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
nt, offsets = jagged_from_list([a, b, c], None)
nt2, _ = jagged_from_list([a, b, c], offsets)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```

Error:
```
(pytorch3) jw3468-mbp:pytorch3 jw3468$ t3
Traceback (most recent call last):
  File "/Users/jw3468/local/.tests/tst3.py", line 932, in <module>
    out = compiled_f(nt, nt2)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 451, in transform
    tracer.run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2103, in run
    super().run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 743, in run
    and self.step()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 706, in step
    getattr(self, inst.opname)(inst)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2213, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 855, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/jw3468/opt/miniconda3/envs/pytorch3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 983, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1050, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1035, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/__init__.py", line 1642, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4541, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

[ghstack-poisoned]

soulitzer committed:
Update on "[ignore] AOTAutograd NestedTensor testing"


Repro:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
nt, offsets = jagged_from_list([a, b, c], None)
nt2, _ = jagged_from_list([a, b, c], offsets)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```

Error:
```
(pytorch3) jw3468-mbp:pytorch3 jw3468$ t3
Traceback (most recent call last):
  File "/Users/jw3468/local/.tests/tst3.py", line 932, in <module>
    out = compiled_f(nt, nt2)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 406, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/eval_frame.py", line 554, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 140, in _fn
    return fn(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 380, in _convert_frame_assert
    return _compile(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 559, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 481, in compile_inner
    out_code = transform_code_object(code, transform)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/bytecode_transformation.py", line 1028, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/convert_frame.py", line 451, in transform
    tracer.run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2103, in run
    super().run()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 743, in run
    and self.step()
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 706, in step
    getattr(self, inst.opname)(inst)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/symbolic_convert.py", line 2213, in RETURN_VALUE
    self.output.compile_subgraph(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 855, in compile_subgraph
    self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
  File "/Users/jw3468/opt/miniconda3/envs/pytorch3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 983, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1050, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/output_graph.py", line 1035, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/jw3468/local/pytorch3/torch/__init__.py", line 1642, in __call__
    return self.compiler_fn(model_, inputs_, **self.kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/backends/common.py", line 55, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4541, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/Users/jw3468/local/pytorch3/torch/_dynamo/utils.py", line 190, in time_wrapper
    r = func(*args, **kwargs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

[ghstack-poisoned]

soulitzer commented:
Currently this PR does not yet work due to the following issue:

Given code:
```python
from torch.nested._internal.nested_tensor import NestedTensor, jagged_from_list, ViewBufferFromNested

a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64)
b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64)
c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64)
# Note that offsets here are not the same tensor
nt, offsets = jagged_from_list([a, b, c], None)
nt2, offsets = jagged_from_list([a, b, c], None)

def fn1(nt1, nt2):
    if nt1.shape[1] == nt2.shape[1]:
        out = nt1 + nt2
        return out
    else:
        out = nt1.sin()
        return out

compiled_f = torch.compile(fn1, fullgraph=True, backend="aot_eager", dynamic=True)
out = compiled_f(nt, nt2)
out_buffer = ViewBufferFromNested.apply(out)
out_buffer.sum().backward()
```
We run into the error:
```
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 4079, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2483, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 2669, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/aot_autograd.py", line 3388, in aot_dispatch_autograd
    fw_module, bw_module = aot_config.partition_fn(
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 693, in min_cut_rematerialization_partition
    orig_fw_outputs, required_fw_nodes, required_bw_nodes, unclaimed_nodes = classify_nodes(joint_module)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 686, in classify_nodes
    forward_only_graph = _extract_graph_with_inputs_outputs(joint_module.graph, inputs, fwd_outputs)
  File "/Users/jw3468/local/pytorch3/torch/_functorch/partitioners.py", line 97, in _extract_graph_with_inputs_outputs
    assert not isinstance(env[x], InvalidNodeBase), f"Node {x} was invalid, but is output"
torch._dynamo.exc.BackendCompilerFailed: backend='aot_eager' raised:
AssertionError: Node tangents_2 was invalid, but is output
```

From the dense -> dense graph traced out by aot autograd, we observe the following:
```
opcode         name        target            args                                                           kwargs
-------------  ----------  ----------------  -------------------------------------------------------------  --------
placeholder    primals_1   primals_1         ()                                                             {}
placeholder    primals_2   primals_2         ()                                                             {}
placeholder    primals_3   primals_3         ()                                                             {}
placeholder    primals_4   primals_4         ()                                                             {}
placeholder    primals_5   primals_5         ()                                                             {}
placeholder    primals_6   primals_6         ()                                                             {}
placeholder    primals_7   primals_7         ()                                                             {}
placeholder    tangents_1  tangents_1        ()                                                             {}
placeholder    tangents_2  tangents_2        ()                                                             {}
call_function  sin         aten.sin.default  (primals_3,)                                                   {}
call_function  cos         aten.cos.default  (primals_3,)                                                   {}
call_function  mul         aten.mul.Tensor   (tangents_1, cos)                                              {}
output         output      output            ([sin, tangents_2, None, None, mul, tangents_2, None, None],)  {}
```
The inputs (placeholders) look mostly as expected: We see 7 primals passed in because 3 are symints, and we have 2 primals for each subclass. We have some None's returned as grad_out probably because we need to save those symints for backward (its interesting that there's a 4th one though). There are 2 tangents as input because there's a single output.

The outputs are more suspicious however, `([sin, tangents_2, None, None, mul, tangents_2, None, None],)`, and this is probably why the partitioner is erroring out. In particular, the partitioner expects the first two outputs to correspond to the forward outputs for the forward graph, but the second output is tangents_2 which would only get passed as an input during backward!

So what I'm suspecting is somewhere in the stack, the input deduping logic has deduplicated the offset corresponding to the subclass in the input and the offset corresponding to the grad_output because they are the same tensor. I'm not sure where that could be happening since I think `aot_wrapper_dedupe `happens on the subclass -> subclass graph.

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer commented:
What was happening in the dense -> dense graph above was that this is just something that naturally happens in fx when you a function like the following:

```
from functorch import make_fx

def f(a, b, c):
    return a, c

x = torch.tensor(1.)
y = torch.tensor(1.)

fx_g = make_fx(f)(x, y, x)
fx_g.graph.print_tabular()
```

That would give the graph:
```
opcode       name    target    args           kwargs
-----------  ------  --------  -------------  --------
placeholder  a_1     a_1       ()             {}
placeholder  b_1     b_1       ()             {}
placeholder  c_1     c_1       ()             {}
output       output  output    ((c_1, c_1),)  {}
```

The reason this is happening is that during trace time, in order to trace out the backward, we need to speculatively produce some kind of grad_out, and way we get that today is by reusing the outputs as is (fake tensors).

In this PR the current workaround that seems to pass a version of the repro above is to make the grad_out a fresh NestedTensor with the offsets cloned, but with the ragged_size preserved.

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1204,6 +1206,14 @@ def inner(*flat_args):
+                ragged_size = x._size[x._ragged_idx] if should_prop else None
+                return NestedTensor(x._values, x._offsets.clone(), ragged_size=ragged_size)
+            else:
+                return x
```
This feels a little weird, but here's how I'm trying to rationalize it.

This feels weird because normally if I clone offsets in user code, I would run into the problem where if I query for the "singleton int" of that offset it would be different. 

In this case, I have some s0 corresponding to a real value (hint) and its sympy counterpart SingletonInt (a subclass of AtomicExpr). That hint corresponds to a particular non-fake offsets, which makes it feel weird, but maybe it doesn't matter, and here's why:

Since everything is symbolic, all I have to work with are fake tensors, and it is not verifiable whether that hint corresponds to any fake tensor. This is the same reason why we propagate ragged_size everywhere anyway. Therefore in the symbolic world, no matter what fake offsets you are holding you are you free to pretend to actually have the "id" of any other offsets.

soulitzer replied:
(this function probably belongs in nested_tensor.py, if we do indeed want something like it)

bdhirsh replied:
Hmm one question - Is it the case that this entire problem came around because of the fact that we're tracing a joint graph where some of the primals and tangents are the **same** tensor (offsets shows up as both `primals_4` and `tangents_2`)? If so, what do you think of the following:

```
def view_avoid_dupes_with_primals(t):
    if isinstance(t, Tensor) and is_traceable_wrapper_subclass(t):
        return transform_subclass(t, lambda _, inner_t: view_avoid_dupes_with_primals(inner_t))
    if isinstance(t, Tensor):
        return t.view(t.shape)
    return t

traced_tangents = pytree.tree_map(view_avoid_dupes_with_primals, traced_tangents)
```

This will just do a (sort of no-op) view on every inner-tensor-tangent to the joint graph, that way we're guaranteed we have no duplicates across the primals and tangents.
(1) no need to special-case NestedTensor (or how that the subclass has a reasonable implementation of clone), since it should always be safe to view the inner tensors of a subclass
(2) I did view here vs. clone - I guess it doesn't matter too much, although clone maybe feels like overkill since all we really care about is tensor identity

soulitzer replied:
Yes that is correct, and adding the more general support sounds good to me! It is also cool that this would support nesting subclasses.

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -129,6 +130,12 @@ def __tensor_unflatten__(inner_tensors: Dict, meta):
         values = inner_tensors["_values"]
         offsets = inner_tensors["_offsets"]
 
+        should_prop = any(any(is_symbolic(s) for s in x.shape) for x in (values, offsets))
```
Why not just check is_fake?

soulitzer replied:
So the reason is that `is_fake(FunctionalTensor())` returns False, but can hold symbolic sizes. I think its holding a fake tensor underneath? but since it's not considered a desugaring subclass, `is_fake` does not understand that.

soulitzer commented on torch/_functorch/aot_autograd.py:
```diff
@@ -31,14 +31,16 @@
 from torch.fx import immutable_collections, Interpreter
 from torch.fx.experimental.proxy_tensor import is_sym_node, py_sym_types
-from torch.fx.experimental.symbolic_shapes import ShapeEnv, is_concrete_int, fx_placeholder_vals
+from torch.fx.experimental.symbolic_shapes import ShapeEnv, is_concrete_int, fx_placeholder_vals, is_symbolic
```
Unnecessary import

bdhirsh commented:
> The reason this is happening is that during trace time, in order to trace out the backward, we need to speculatively produce some kind of grad_out, and way we get that today is by reusing the outputs as-is! (That means that, in my dense graph, if I returned an input as-is as one of my outputs, I would end up having two inputs that are the same.)

> In this PR the current workaround that seems to pass a version of the repro above is to make the grad_out a fresh NestedTensor with the offsets cloned. By cloning, we would no longer run into the fx problem above.

> One detail to take note of is that we need to potentially propagate the ragged_size of the output.

Really great debugging!

pre-subclass-world, we managed to avoid this because we explicitly do NOT create grad_outputs during tracing for any outputs that alias inputs (filtered out here: [code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1218)). But if a subclass does extra aliasing between graph outputs/inputs in a way that is not visible above the sublass, then we can end up having tangents that alias inputs like you said.

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3696,6 +3696,10 @@ def _maybe_evaluate_static(
         new_shape_env = {}
         new_range_env = {}
         for idx, k in enumerate(symbols):
+            if isinstance(self.var_to_val.get(k, None), SingletonInt):
```
@ezyang should probably take a look?

ezyang replied:
this LGTM. I didn't check the rest of the PR

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



[ghstack-poisoned]

soulitzer commented:
Updates:
- Generalized the workaround mechanism for https://github.com/pytorch/pytorch/pull/110529#issuecomment-1758616409  per https://github.com/pytorch/pytorch/pull/110529#discussion_r1357669300
- Add test (only for nested tensor so far, but let me know if there's a more general subclass case I should test)
- Added note about holding symbolic attributes on subclasses

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1204,6 +1205,16 @@ def inner(*flat_args):
 
+        # If a subclass does extra aliasing between graph outputs/inputs in a way
+        # that is not visible above the sublass, then we can end up having tangents
+        # that alias inputs.
```
nit: Can you add more details to this note?

Ordinarily, when constructing the joint function that we want to trace in AOTAutograd, we're guaranteed that we the tangent tensors that we pass into the joint are distinct tensors from the primals. This is because when decide which forward outputs to create tangents for, we **only** create tangents for forward outputs that are not aliases of inputs (See Note [AOT Autograd: outputs aliasing inputs or intermediates!]).

However, when wrapper tensor subclasses enter the picture, it is possible to have an output of the forward that is a subclass that is not an input / alias of an input, **but** one of its inner tensors is an alias! NestedTensor is an example: Performing an out-of-place pointwise op on a NestedTensor will often construct a fresh NestedTensor that holds onto the input's offsets tensor directly.

Ordinarily, we can just pass in the traced (fake tensor) fw outs directly as our tangents when tracing the joint. But if we have tangent tensors that **are** the same as the (primal) forward inputs, this can cause problems during tracing: make_fx() will specialize on our duplicate inputs. If we passed in the same tensor for `primals_1` and `tangents_1` during tracing, `make_fx()` will happily sub out all usages of `tangents_1` with `primals_1` in the graph, which is not what we want.

To work around this, we view every forward output when creating out tangent tensors.

soulitzer replied:
Thanks for the detailed write up! I've included the details into the note and moved it to the top of the file with the other notes to avoid breaking the flow of the code since its a little longer now (let me know if you prefer it inline).

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented on test/dynamo/test_subclasses.py:
```diff
@@ -633,6 +637,39 @@ def test_binary_recompiles_due_to_duck_sizing(self):
+        self._test_autograd("aot_eager")
+
+    @requires_cuda()
+    def test_basic_autograd_inductor(self):
```
this test is specified to NestedTensor right? Mention that in the test name?

soulitzer replied:
Oh this one is already in the TestNestedTensor test class

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3620,7 +3660,8 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
             forward_saved_for_backwards_strides = None
             if fwd_output_strides is not None:
-                forward_saved_for_backwards_strides = fwd_output_strides[fw_metadata.tensors_saved_for_backwards_slice]
+                inner_meta.num_symints_saved_for_bw = fw_metadata.num_symints_saved_for_bw
```
nit: I would move this up to the same place where `fw_metadata.mum_symints_saved_for_bw` is also set, for clarity.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3913,11 +3951,25 @@ def backward(ctx, *flat_args):
-
             # TODO: figure out how to refactor the backward properly so I can use aot_dispatch_subclass_wrapper() here.
+            tangents_start_idx = len(all_args) - num_flat_bw_args_with_grads - len(rng_args)
+            tangents_end_idx = len(all_args) - len(rng_args)
```
I would also move these index calculations to right where `all_args` is defined too, for clarity (so it's super clear why you computed it this way).

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3913,11 +3951,25 @@ def backward(ctx, *flat_args):
+            # Make the tangents contiguous. Note that we must do this after subclass desugaring
+            # because inputs to inductor have to be contiguous
+            all_args = [
+                t.contiguous() if isinstance(t, torch.Tensor) and tangents_start_idx <= i < tangents_end_idx else t
```
hmm, is the `isinstance(t, torch.Tensor)` check necessary here? I might be wrong but I thought that all tangents have to be tensors (we won't e.g. have symints here, since symints will be directly saved in the ctx).

soulitzer replied:
Ah I wasn't sure because this was preexisting and I was worried about Nones from undefined gradients, but I think we need to materialize them today, so I think you're right.

bdhirsh reviewed (approved):
nice work! Left a few more nits but mostly looks good to me.

soulitzer committed:
Update on "Compile NestedTensor with AOTAutograd"


This PR has a number of changes that improve subclass support for AOTAutograd/Inductor in general:
-  previously if a subclass does extra aliasing between graph outputs/inputs in a way, the partitioner would complain because grad_outputs are the outputs reused as-is. Now we do a view_as(self) to workaround this.
- Use dense -> dense metadata when working with fwd_output_strides during backward. This is important since the stride information comes from inductor which sees the dense to dense graph.
- Inductor requires that the inputs to the compiled backward to match some expected strides computed during compilation. We make sure to make the inner tensors of the subclass contiguous (previously, we only made the subclass itself contiguous)

Changes specific to NestedTensor relevant to compilation:
- Properly handle the case where `__tensor_unflatten__` is passed non-symbolic dense tensors and with meta extracted from fake subclasses.
- Skip var_to_range logic for singleton int
- Skip size hint logic in inductor for singleton int

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge
